File: Ceasar/Ceasar.cpp
#include <iostream>
using namespace std;

void Annotation(string text, int key, bool encrypt);
int main()
{
    string text;
    int key;
    bool encrypt;
    Annotation(text, key, encrypt);
    return 0;
}
// * Function to shift a character by key value
char shiftChar(char inputChar, char conditionChar, int key, bool encrypt = true)
{
    if (encrypt)
        inputChar = (inputChar - conditionChar + key) % 26 + conditionChar;
    else
        inputChar = (inputChar - conditionChar - key + 26) % 26 + conditionChar;
    return inputChar;
}
// * Function to shift a string by key value
string shiftString(string text, int key, bool encrypt = true)
{
    string result = "";
    for (int i = 0; i < text.length(); i++)
    {
        if (isupper(text[i]))

            result += shiftChar(text[i], 'A', key, encrypt);

        else if (islower(text[i]))

            result += shiftChar(text[i], 'a', key, encrypt);

        else

            result += text[i];
    }
    return result;
}

string Ceasar(string text, int key, bool encrypt = true)
{
    return shiftString(text, key, encrypt);
}
// * Function to take input from user and give output
void Annotation(string text, int key, bool encrypt)
{
    try
    {
        cout << "Enter the text to be encrypted: ";
        getline(cin, text);
        cout << "Enter the key value: ";
        cin >> key;
        if (cin.fail())
        {
            throw invalid_argument("Invalid key value. Please enter an integer.");
        }
        cout << "Encrypt or Decrypt? (1/0): ";
        cin >> encrypt;
        if (cin.fail())
        {
            throw invalid_argument("Invalid choice for encrypt/decrypt. Please enter 1 or 0.");
        }
    }
    catch (const invalid_argument &e)
    {
        cerr << "Error: " << e.what() << endl;
        exit(1);
    }
    cout << "The result text is: " << Ceasar(text, key, encrypt) << endl;
}

File: Ceasar/Ceasar.py
"""
*Change characters in a string by shifting them by a fixed amount
@param text: string to be encrypted/decrypted, @param key: shift amount, @param encrypt: boolean to encrypt or decrypt
"""
def shiftChar(char, key, encrypt=True):
    if char.isupper():
        base = 65
    elif char.islower():
        base = 97
    else:
        return char
    
    if encrypt:
        return chr((ord(char) + key - base) % 26 + base)
    return chr((ord(char) - key - base) % 26 + base)
"""
*Encrypt or decrypt a string using the Ceasar cipher
@param text: string to be encrypted/decrypted, @param key: shift amount, @param encrypt: boolean to encrypt or decrypt
""" 
def Ceasar(text, key, encrypt=True):
    result = ""
    for i in range(len(text)):
        result += shiftChar(text[i], key, encrypt)
    return result
"""
*Encrypt or decrypt a string using the Ceasar cipher with recursion
@param result: the string to be returned, @param currentCharIndex: the index of the current character, @param text: the text to be encrypted/decrypted, @param key: the shift amount
""" 
def CeaserRecursive(result,currentCharIndex , text, key, encrypt=True):
    if currentCharIndex >= len(text):
        return result
    result += shiftChar(text[currentCharIndex], key, encrypt)
    return CeaserRecursive(result, currentCharIndex+1, text, key,encrypt)

def main():
    try:
        text = input("Enter the text: ")
        key = int(input("Enter the key: "))
        encrypted = bool(int(input("Encrypt or Decrypt? (1/0): ")))
    except ValueError:
        print("Invalid input. Please enter valid text, a numeric key, and 1 or 0 for encryption/decryption.")
        exit(1)

    # print("Result: ", Ceasar(text, key, encrypted))
    print("Recursive Result: ", CeaserRecursive("", 0, text, key, encrypted))

main()